frontend/app/components/work-packages/work-package-cache.service.ts

```diff
@@ -50,31 +52,49 @@ export class WorkPackageCacheService {
   }
 
   updateWorkPackage(wp: WorkPackageResource) {
-    this.updateWorkPackageList([wp]);
+    // this.updateWorkPackageList([wp]);
+    states.workPackages.put(wp.id.toString(), wp);
   }
 
   updateWorkPackageList(list: WorkPackageResource[]) {
     for (const wp of list) {
-      var cached = this.workPackageCache[wp.id];
-      if (cached && cached.dirty) {
-        this.workPackageCache[wp.id] = cached;
-      } else {
-        this.workPackageCache[wp.id] = wp;
-      }
+      // var cached = this.workPackageCache[wp.id];
+
+      // TODO RR Question: This check doesn't seem to be right!?
+      // if (cached && cached.dirty) {
+      //   this.workPackageCache[wp.id] = cached;
+      // } else {
+      //   this.workPackageCache[wp.id] = wp;
+      // }
+
+      states.workPackages.put(wp.id.toString(), wp);
     }
-    this.workPackagesSubject.onNext(this.workPackageCache);
+    // this.workPackagesSubject.onNext(this.workPackageCache);
   }
 
-  loadWorkPackage(workPackageId: number, forceUpdate = false): Rx.Observable<WorkPackageResource> {
-    if (forceUpdate || this.workPackageCache[workPackageId] === undefined) {
-      this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate).then(wp => {
-        this.updateWorkPackage(wp);
-      });
+  loadWorkPackage(workPackageId: number, forceUpdate = false): State<WorkPackageResource> {
+    // if (forceUpdate || this.workPackageCache[workPackageId] === undefined) {
+    //   this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate).then(wp => {
+    //     this.updateWorkPackage(wp);
+    //   });
+    // }
+    //
+    // return this.workPackagesSubject
+    //   .map(cache => cache[workPackageId])
+    //   .filter(wp => wp !== undefined);
+
+    const state = states.workPackages.get(workPackageId.toString());
+    if (forceUpdate) {
+      state.clear();
+      // this.apiWorkPackages.loadWorkPackageById(workPackageId, forceUpdate).then(wp => {
+      //   state.put(wp);
+      // });
     }
 
-    return this.workPackagesSubject
-      .map(cache => cache[workPackageId])
-      .filter(wp => wp !== undefined);
+    state.putFromPromiseIfPristine(
```

Comment:
@oliverguenther Despite our discussion, I added this util method. This case seems to be so common that I think it makes sense to add this method. Also, the logic/intention of this method seems to be more obvious this way. Ok for you?